### DISCUSSION TITLE
Supporting Physical Interface Devices for Gamepad with Force Feedback

### DISCUSSION THREAD

**Discussion by Tails86** [hathach/tinyusb#1670] [Q&A] [answered]
I've been diving into [Device Class Definition for Physical Interface Devices](https://usb.org/sites/default/files/pid1_01_0.pdf) in order to support force feedback (vibration) for a gamepad device I've been developing. I have begun by defining Table 2 in the linked document as an enum since I don't see that defined as part of tinyusb library (please correct me if I'm wrong). Now I'm trying to extend the existing TUD_HID_REPORT_DESC_GAMEPAD macro to include force feedback definitions.

I'm mainly here looking for any guidance or examples of how to implement a gamepad with force feedback. The best example I have found so far is [here](https://www.microchip.com/forums/m958105.aspx) which seems to follow section 9 of the PID document. I find it a little hard to follow because some externally referenced data/code is missing, but it has been very helpful seeing some code which is claimed to have worked, even though it is using a different USB device library.

The first thing I have run into is that all examples I have run into use "joystick" instead of "gamepad". As far as I can tell from [HID Usages and Descriptions](https://usb.org/sites/default/files/hut1_3_0.pdf), joystick seems to be interchangeable with gamepad except for the following difference:

> Traditionally, a joystick driver applies its own scaling to values returned from a joystick. That is, the driver simply linearizes and translates the range of values generated by the stick into normalized values between 0 and 64K, where 32K is centered. The application (game) then interprets the normalized values as necessary. Because of this, joysticks normally do not declare Units or Physical Minimum and Physical Maximum values for their axes. Depending on the driver, these items may be ignored if they are declared.

Should I interpret that to mean joystick analog sticks are internally calibrated by the device and gamepad analog sticks are externally calibrated by the host? Are these samples then really interchangeable with gamepad besides that difference? 

Has there ever been any effort to provide standard reports/structs to support vibration in tinyusb? If not, would contributing back to this project be seen as useful?

**Comment by hathach** [comment, marked as answer]
> Has there ever been any effort to provide standard reports/structs to support vibration in tinyusb? If not, would contributing back to this project be seen as useful?

I haven't worked with this feature and currently don't have any schedule to provide support for vibration. For hints, you probably look at other project and/or reverse engineer report descriptor from an consumer devices. Any contributing back is welcome, it will help others with similar problems, also making improving/enhancing new feature easier.
